src/webgl/context-state.js

```diff
@@ -0,0 +1,540 @@
+    webgl2: true
+  },
+  //  Number of pixel images skipped before first pixel is read from memory
+  unpackSkipPixels: {
+    params: GL.UNPACK_SKIP_PIXELS,
+    setter: (gl, value) => gl.pixelStorei(value),
+    webgl2: true
+  },
+  //  Number of rows of pixels skipped before first pixel is read from memory
+  unpackSkipRows: {
+    params: GL.UNPACK_SKIP_ROWS,
+    setter: (gl, value) => gl.pixelStorei(value),
+    webgl2: true
+  },
+  //  Number of pixel images skipped before first pixel is read from memory
+  unpackSkipImages: {
+    params: GL.UNPACK_SKIP_IMAGES,
+    setter: (gl, value) => gl.pixelStorei(value),
+    webgl2: true
+  }
+};
+
+// Map from GL parameter constants to composite paramters
+const GL_PARAMS = {};
+
+function unpackStateParams() {
+  for (const key in GL_STATE) {
+    const parameterDef = GL_STATE[key];
+    const {params} = parameterDef;
+    const paramsArray = isArray(params) ? params : [params];
+    for (const param of paramsArray) {
+      GL_PARAMS[param] = key;
+    }
+  }
+}
+
+unpackStateParams();
+
+// GETTERS AND SETTERS
+
+/**
+ * Sets value with key to context.
+ * Value may be "normalized" (in case a short form is supported). In that case
+ * the normalized value is retured.
+ *
+ * @param {WebGLRenderingContext} gl - context
+ * @param {String} key - parameter name
+ * @param {*} value - parameter value
+ * @return {*} - "normalized" parameter value after assignment
+ */
+export function getGLParameter(gl, key) {
+  const parameterDefinition = GL_STATE[key];
+  if (!parameterDefinition) {
+    throw new Error(`Unknown GL state parameter ${key}`);
+  }
+  // Get the parameter value(s) from the context
+  const {params} = parameterDefinition;
+  const value = isArray(params) ?
+    params.map(param => gl.getParameter(param)) :
+    gl.getParameter(params);
+  return value;
+}
+
+/**
+ * Sets value with key to context.
+ * Value may be "normalized" (in case a short form is supported). In that case
+ * the normalized value is retured.
+ *
+ * @param {WebGLRenderingContext} gl - context
+ * @param {String} key - parameter name
+ * @param {*} value - parameter value
+ * @return {*} - "normalized" parameter value after assignment
+ */
+export function setGLParameter(gl, key, value) {
+  const parameterDefinition = GL_STATE[key];
+  if (!parameterDefinition) {
+    throw new Error(`Unknown GL state parameter ${key}`);
+  }
+  const {setter, normalizeValue} = parameterDefinition;
+  const adjustedValue = normalizeValue ? normalizeValue(value) : value;
+  setter(gl, adjustedValue);
+  return adjustedValue;
+}
+
+// HELPERS
+
+function isArray(array) {
+  return Array.isArray(array) || ArrayBuffer.isView(array);
+}
+
+// GLState
+
+class GLState {
+  // Note: does not maintain a gl reference
+  constructor(gl, {copyState = false} = {}) {
+    this.state = {};
+    if (copyState) {
+      this._copyWebGLState(gl);
+    } else {
+      this._getInitialState();
+    }
+    this.overrides = [];
```

Comment:
I feel that the name "overrides" is not intuitive enough. Can we call it stateStack? or similar?